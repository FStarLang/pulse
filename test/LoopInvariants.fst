module LoopInvariants

#lang-pulse
open Pulse

fn test0 ()
{
  while (true)
    invariant (emp)
  { (); }
}

fn test1 ()
{
  (* This failed to parse, confusing the opening brace
  for the start of a refinement. `invariant (emp)` worked*)
  while (true)
    invariant emp
  { (); }
}

(* Multiple invariant annotations in a loop are just
starred together. *)
fn test2 (r s : ref int)
  preserves r |-> 1
  preserves s |-> 2
{
  while (true)
    invariant r |-> 1
    invariant emp
    invariant s |-> 2
  { (); }
}

(* You can have no invariants, that's the same as annotating
emp. *)
fn test3 (r s : ref int)
  preserves r |-> 1
  preserves s |-> 2
{
  while (true)
  {
    r := !s - 1;
    s := !r + 1;
  }
}

fn test4 ()
{
  while (true)
    invariant emp
    invariant emp
  { (); }
}

(* One cannot mix old-style and new-style invariants. *)
[@@expect_failure]
fn test5 (r s : ref int)
  preserves r |-> 1
  preserves s |-> 2
{
  while (true)
    invariant r |-> 1
    invariant b. emp
  { (); }
}

[@@expect_failure]
fn test6 (r s : ref int)
  preserves r |-> 1
  preserves s |-> 2
{
  while (true)
    invariant r |-> 1
    invariant emp
    invariant s |-> 2
  { s := 3; }
}

[@@expect_failure]
fn test7 (r s : ref int)
  preserves r |-> 1
  preserves s |-> 2
{
  (* We infer the same invariant as above. *)
  while (true)
    invariant r |-> 1
  { s := 3; }
}
