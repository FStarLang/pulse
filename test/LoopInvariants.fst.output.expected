* Info at LoopInvariants.fst(64,17-64,20):
  - Expected failure:
  - When using multiple invariants, they must all be in the "new" style without a binder.

* Info at LoopInvariants.fst(76,20-77,10):
  - Expected failure:
  - rewrite: could not prove equality of
  - Pulse.Lib.Reference.pts_to s 3
  - Pulse.Lib.Reference.pts_to s 2
  - Assertion failed
  - The SMT solver could not prove the query. Use --query_stats for more
    details.

* Info at LoopInvariants.fst(83,18-88,10):
  - Expected failure:
  - rewrite: could not prove equality of
  - Pulse.Lib.Reference.pts_to s 3
  - Pulse.Lib.Reference.pts_to s 2
  - Assertion failed
  - The SMT solver could not prove the query. Use --query_stats for more
    details.

